cpp/ql/src/semmle/code/cpp/dataflow/TaintTracking.qll

```diff
@@ -231,6 +232,31 @@ module TaintTracking {
     e instanceof AlignofOperator
   }
 
+  private predicate exprToExprStep(Expr exprIn, Expr exprOut) {
+    exists(Call call, TaintFunction f, FunctionInput inModel, FunctionOutput outModel |
+      call.getTarget() = f and
+      f.hasTaintFlow(inModel, outModel) and
+      (
+        exists(int argInIndex |
+          inModel.isInParameterPointer(argInIndex) and
+          exprIn = call.getArgument(argInIndex)
+        ) or exists(int argInIndex |
+          inModel.isInParameterPointer(argInIndex) and
+          call.passesByReference(argInIndex, exprIn)
```

Comment:
The same pattern occurs in `exprToDefinitionByReferenceStep` in master.  Is that wrong?  Or is there a difference?

(I should probably add test coverage for this specific line if we don't decide to get rid of it...)

Reply:
It's actually wrong what I said: `passesByReference` holds also for references to non-const.

Reply:
I think the pattern is also wrong in `TaintTrackingUtil.exprToDefinitionByReferenceStep`. Since this is taint tracking, data will flow through `&` already, so that part of `passesByReference` should not contribute anything. This pattern will allow flow from `x` to `&x.a` in the context of `f(&x.a)`, but is that the intended effect? Hopefully that's handled by field flow; if not, please write a test case and let me know.